### PR TITLE
Support CgroupsV2 in info.cpp's calculations

### PR DESCRIPTION
**Pull request by elukey**
There is currently support for CPUs count approximation for Cgroups V1, but not for Cgroups V2.
This change adds a similar heuristic as
https://github.com/dmlc/xgboost/issues/9622, so
that the count doesn't fall back to /proc/cpuinfo
on more recent systems.

issue: #2518

Before submitting a pull request, please do the following steps:

1. Read [instructions for contributors](https://catboost.ai/docs/concepts/development-and-contributions.html).
2. Run `ya make` in catboost folder to make sure the code builds.
3. Add tests that test your change.
4. Run tests using `ya make -t -A` command.
5. If you haven't already, complete the [CLA](https://yandex.ru/legal/cla/).

### CODE REVIEW

**Comment by elukey** [comment]
Still working on building the code locally, I'd also like to get some info about how to better test it :)

**Comment by elukey** [comment]
I am able to build the Python wheel via `python3 setup.py bdist_wheel`. Need to find a minimal test to verify the new code.

**Comment by andrey-khropov** [comment]
Please, agree to the [CLA](https://yandex.ru/legal/cla/).

**Comment by andrey-khropov** [comment]
> I am able to build the Python wheel via `python3 setup.py bdist_wheel`. Need to find a minimal test to verify the new code.

See [the documentation about the tests](https://catboost.ai/en/docs/concepts/development-and-contributions#cmake-based-build-tests)

**Comment by elukey** [comment]
@andrey-khropov Hi! Thanks a lot for the feedback! I have a couple of questions:
1) I am using the python wrapper to build with cmake, but from the test docs it seems that I'd need to run cmake directly to build a specific test. What would you use to build `utils/system/info_ut.cpp` ? I get what the target should be (`util-system-info-ut`) but I can't make cmake work :(
2) To properly test the code I'd need to mock some sysfs files, or to somehow create tmp files with the content that I want and inject them in the code (in this case, in static/internal methods). I am not 100% familiar with how to do this in cpp/catboost, do you have any example/recommendation?

**Comment by elukey** [comment]
I am still trying to figure out how to run tests (reading https://catboost.ai/en/docs/concepts/development-and-contributions#run-tests), any suggestion?

**Comment by andrey-khropov** [comment]
> @andrey-khropov Hi! Thanks a lot for the feedback! I have a couple of questions:
> 
>     1. I am using the python wrapper to build with cmake, but from the test docs it seems that I'd need to run cmake directly to build a specific test. What would you use to build `utils/system/info_ut.cpp` ? I get what the target should be (`util-system-info-ut`) but I can't make cmake work :(
> 
>     2. To properly test the code I'd need to mock some sysfs files, or to somehow create tmp files with the content that I want and inject them in the code (in this case, in static/internal methods). I am not 100% familiar with how to do this in cpp/catboost, do you have any example/recommendation?

Somehow CMake build for unittests in `catboost/util` in particular seems to be non-working. I'll look into it: #2525 . I'm sorry.

**Comment by elukey** [comment]
Thanks! I'll concentrate on finding a good unit test for the new code :)

**Comment by elukey** [comment]
I don't have a better idea to test this than to run a special unit test that prints the number of cpus with different container specs. It will be a one-off, but it is inline with the current tests for util/info. If there is a way to mock some sysfs file contents in the unit tests we may be able to have something better, but I don't have any expertise in c++ testing so I'd need some guidance.

**Comment by elukey** [comment]
Built the static library of catboost and used the following to test cgroups v2:

```
#include "util/system/info.h"
#include "stdio.h"

int main() {
    long int cpus = NSystemInfo::NumberOfCpus();
    printf("CPUs: %ld\n", cpus);
}
```

On Debian Bookworm (default is using cgroups v2) I tested:

* docker run --cpus=2  -it -v $(pwd):/test debian:bookworm (CPUS: 2)
* docker run --cpus=0.5  -it -v $(pwd):/test debian:bookworm (CPUS: 1)
* docker run -it -v $(pwd):/test debian:bookworm (CPUS: 8, max present on my laptop)

**Comment by elukey** [comment]
I agree with the CLA outlined in https://yandex.ru/legal/cla/?lang=en

File: util/system/info.cpp [hunk 1]
```diff
@@ -68,6 +69,48 @@ static inline double CgroupCpus() {
...
+
+static inline double CgroupCpus() {
+    auto cgroup2_cpus = CgroupV2Cpus("/sys/fs/cgroup/cpu.max");
+    if (cgroup2_cpus > 0) {
+        return cgroup2_cpus;
+    }
+    auto cgroups1_cpus = CgroupV1Cpus(
+        "/sys/fs/cgroup/cpu/cpu.cfs_quota_us",
+        "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
+    );
```
Comment (andrey-khropov) [inline util/system/info.cpp:108]: Please put ");" on the previous line, our internal style check for `util` enforces this style.

```
E     --- L
E     +++ R
E     @@ -104,8 +104,7 @@
E          }
E          auto cgroups1_cpus = CgroupV1Cpus(
E              "/sys/fs/cgroup/cpu/cpu.cfs_quota_us",
E     -        "/sys/fs/cgroup/cpu/cpu.cfs_period_us"
E     -    );
E     +        "/sys/fs/cgroup/cpu/cpu.cfs_period_us");
E          if (cgroups1_cpus > 0) {
E              return cgroups1_cpus;
E          }
```

**Review by andrey-khropov — CHANGES_REQUESTED** [review]
Please, change the names of the local variables to `camelCase` style according to the  [C++ style guide](https://github.com/catboost/catboost/blob/master/CPP_STYLE_GUIDE.md#variables)

File: util/system/info.cpp (hunk 1 above)
Comment (elukey) [reply util/system/info.cpp]: Done!

**Comment by elukey** [comment]
@andrey-khropov thanks! Sent a new patch with variables in camel case and the ) on the previous line. Re-tested as explained above as well.

**Review by andrey-khropov — APPROVED** [review]

**Comment by elukey** [comment]
@andrey-khropov thanks a lot for the approval, what is the next step? Should I help/work on https://github.com/catboost/catboost/issues/2525 ?

**Comment by elukey** [comment]
Gentle ping, not sure what to do folks :)

**Comment by andrey-khropov** [comment]
> Gentle ping, not sure what to do folks :)

I've discovered some strange buffer underflow issue when I tried to merge this PR but it is not related to your changes directly, I'm investigating. You don't have to do anything for now.

**Comment by elukey** [comment]
> > Gentle ping, not sure what to do folks :)
> 
> I've discovered some strange buffer underflow issue when I tried to merge this PR but it is not related to your changes directly, I'm investigating. You don't have to do anything for now.

Thanks a lot! I asked since I thought the code may have needed more reviewers, and where I work (Wikimedia foundation) we are trying to solve a k8s performance bug when running catboost with cgroups v2. Nothing urgent but ideally we'll love to work with you to have a new minor release in Pypi with the patch to fix the issue :)

**Comment by andrey-khropov** [comment]
> > > Gentle ping, not sure what to do folks :)
> > 
> > 
> > I've discovered some strange buffer underflow issue when I tried to merge this PR but it is not related to your changes directly, I'm investigating. You don't have to do anything for now.
> 
> Thanks a lot! I asked since I thought the code may have needed more reviewers, and where I work (Wikimedia foundation) we are trying to solve a k8s performance bug when running catboost with cgroups v2. Nothing urgent but ideally we'll love to work with you to have a new minor release in Pypi with the patch to fix the issue :)

Yes, CatBoost 1.2.3 that is expected to contain this change is planned to be released this week.

**Comment by elukey** [comment]
@andrey-khropov Hi! Is there anything still pending or shall we merge this? Thanks in advance :)

**Pull request closed without merging**